README.md

[![Travis build
status](https://travis-ci.com/muschellij2/limmi.svg?branch=master)](https://travis-ci.com/muschellij2/limmi)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/muschellij2/limmi?branch=master&svg=true)](https://ci.appveyor.com/project/muschellij2/limmi)
[![Coverage
status](https://codecov.io/gh/muschellij2/limmi/branch/master/graph/badge.svg)](https://codecov.io/gh/muschellij2/limmi)
<!-- README.md is generated from README.Rmd. Please edit that file -->

# limmi Package:

The goal of `limmi` is to provide provides functions for running LIMMA
against medical images.

## Installation

You can install `limmi` from GitHub with:

``` r
# install.packages("remotes")
remotes::install_github("muschellij2/limmi")
```


NEWS.md
# limmi 0.1.0

* Added `nifti_*`, where we currently have `eBayes` and `lmFit`.

* Added a `NEWS.md` file to track changes to the package.

* Added all the functions and the ability to download data.


tests/testthat.R
library(testthat)
library(limmi)

test_check("limmi")


tests/testthat/test-eBayes.R
testthat::context("Whole Pipeline")

tarfile = system.file("extdata", "can.tar.gz", package = "limmi")
tarfile
exdir = tempfile()
dir.create(exdir)
files = untar(tarfile = tarfile, list = TRUE, exdir = exdir)
files = files[!grepl("^\\.", basename(files))]
unz = untar(tarfile = tarfile, files = files, exdir = exdir)
files = files[ grepl("hdr", basename(files))]
files
files = file.path(exdir, files)

testthat::test_that("eBayes works", {
  eb = limmi::nifti_eBayes(files)
  vals = head(na.omit(eb$lm_fit$coefficients[,1]), 10)

  testthat::expect_equal(
    vals,
    c(0.121366907221576, 0.127145683858544, 0.158761932048947, 0.1563262026757,
      0.0135644840387007, 0.0429873446313044, 0.0688248955023786, 0.0809315139971053,
      0.0801648885632555, 0.121430590748787)
  )
})


R/limma_2.R
library(RNifti)
library(limma)


limma_images = function(
  imgs, mask, verbose = TRUE,
  ...,
  proportion = 0.01, stdev.coef.lim = c(0.1,4),
  trend = FALSE, robust = FALSE, winsor.tail.p = c(0.05,0.1),
  adjust.method = "BH") {

  img = asNifti(imgs[[1]])
  if (!is.null(mask)) {
    mask = asNifti(mask) > 0
  } else {
    mask = array(TRUE, dim = dim(img))
  }

  if (packageVersion("neurobase") >= package_version("1.29.1")) {
    mat = neurobase::images_to_matrix(imgs, mask, verbose = verbose)
  } else {
    mat = neurobase::images2matrix(imgs, mask)
  }
  n = nrow(mat)
  fit = limma::lmFit(mat, ...)
  eb.fit = limma::eBayes(
    fit,
    proportion = proportion,
    stdev.coef.lim = stdev.coef.lim,
    trend = trend, robust = robust,
    winsor.tail.p = winsor.tail.p)

  cols_to_grab = c("coefficients", "stdev.unscaled",
                   "t", "p.value", "lods", "adjusted_p_value")
  eb.fit$adjusted_p_value = limma::topTable(
    eb.fit,
    sort.by = "none",
    adjust.method = adjust.method,
    number = n)$adj.P.Val
  out_images = lapply(cols_to_grab, function(x) {
    remake_nifti_image(eb.fit[[x]], img = img, mask = mask)
  })
  names(out_images) = cols_to_grab

  L = list(
    images = out_images,
    empirical_bayes = eb.fit,
    adjust.method = adjust.method,
    lm_fit = fit)
  L
}



R/image_to_SE.R
#' Convert NIfTI data.frame to a SummarizedExperiment
#'
#' @param x An optional `data.frame` describing the samples/participants.
#' @param path_column column in `x` that is the paths to the file
#' names of the data.  Can also be a list column of `x` where this column
#' are the images
#' @param template_images List of images that are annotations of the rows
#' of the data (voxels), such as template mask or segmentations
#' @param assay_name Name of the `assay` in the
#' \code{\link{SummarizedExperiment}}
#' @param ... additional arguments to pass to \code{\link{SummarizedExperiment}}
#'
#' @return A \code{SummarizedExperiment} output
#' @export
#'
#' @examples
#' tarfile = system.file("extdata", "can.tar.gz", package = "limmi")
#' tarfile
#' exdir = tempfile()
#' dir.create(exdir)
#' files = untar(tarfile = tarfile, list = TRUE, exdir = exdir)
#' files = files[!grepl("^[.]", basename(files))]
#' unz = untar(tarfile = tarfile, files = files, exdir = exdir)
#' files = files[ grepl("hdr", basename(files))]
#' files
#' files = file.path(exdir, files)
#' df = data.frame(image = files,
#' age = stats::rpois(length(files), 50))
#' se = nifti_df_to_SummarizedExperiment(df, "image")
nifti_df_to_SummarizedExperiment = function(
  x, path_column = "image",
  assay_name = path_column,
  template_images = NULL,
  ...) {

  imgs = x[[path_column]]
  if (is.factor(imgs)) {
    imgs = as.character(imgs)
  }
  mat = nifti_images_to_matrix(imgs)
  if (!is.null(template_images)) {
    n_timgs = names(template_images)
    if (is.null(n_timgs) & is.character(template_images)) {
      n_timgs = template_images
    } else {
      n_timgs = NULL
    }
    rowData = nifti_images_to_matrix(template_images)
    colnames(rowData) = n_timgs
  } else {
    rowData = NULL
  }
  mat = list(mat)
  if (!is.null(assay_name)) {
    names(mat) = assay_name
  }
  SummarizedExperiment::SummarizedExperiment(
    assays = mat,
    rowData = rowData,
    colData = x,
    ...)
}



R/findNearestNeighbors.R
#' @title Find the approximate set of neighbors of each voxels.
#' @description Find the approximate set of neighhorhood of each voxels,
#' given the distance radius.The number of elements in each neighborhood is (2*radius)^3.
#' We recommend to find the nn prior to limma
#'
#' @param maskImg Input Masked image which the masked voxels are labeled with zeros
#' @param radius Integer: Indices that dist(targetInd - queryInd) <= radius
#' @param threads Integer: threads to use in closest point search.

#' @return The result from \code{\link{vcgKDtree}}
#' @export

findNearestNeighbors = function(maskImg, radius = 1, threads = parallel::detectCores()){
  ind.mask <- which(maskImg > 0, arr.ind = TRUE)
  nn = Rvcg::vcgKDtree(target = ind.mask, query = ind.mask,
                       k = radius^3*8, threads = threads)
  return(nn)
}


R/nifti_images_to_matrix.R
#' @title Transform set of images to matrix
#' @description Creates a matrix, where the voxels are on the rows
#' and images are on the columns
#'
#' @param imgs Vector of files or list of images (niftiImage, array, or nifti)
#' @param mask Binary image to subset the voxels
#' @param verbose print diagnostic messages
#'
#' @return Matrix of V by n, where V is the product of the dimensions of one
#' image or the number of voxels in the mask, and n is the number of images
#' @export 
nifti_images_to_matrix = function(imgs, mask = NULL, verbose = TRUE) {
  if (is.factor(imgs)) {
    imgs = as.character(imgs)
  }
  img = RNifti::asNifti(imgs[[1]])
  if (!is.null(mask)) {
    mask = RNifti::asNifti(mask)
    mask = mask > 0
  } else {
    mask = array(TRUE, dim = dim(img))
  }
  dimg = dim(img)
  pdim = RNifti::pixdim(img)
  rm(img)
  n_voxels_dim = prod(dimg)
  n_voxels = sum(mask, na.rm = TRUE)
  ind = which(mask, arr.ind = TRUE)
  rm(mask)
  n = length(imgs)
  mat = matrix(NA_real_, nrow = n_voxels, ncol = n)
  if (verbose) {
    pb = utils::txtProgressBar(max = n, style = 3)
  }
  for (i in seq(n)) {
    if (verbose) {
      utils::setTxtProgressBar(pb, value = i)
    }
    x = RNifti::asNifti(imgs[[i]])
    stopifnot(prod(dim(x)) == n_voxels_dim)
    mat[,i] = c(x[ind])
  }
  if (verbose) {
    close(pb)
  }
  if (is.character(imgs)) {
    colnames(mat) = imgs
  }
  attr(mat, "image_dim") = dimg
  attr(mat, "image_pixdim") = pdim
  mat
}


R/nifti_local_moderated_t_stat.R
#' Compute local moderated T-statistics
#'
#' @param imgs List of images or character of filenames
#' @param mask Image to subset the data
#' @param verbose Print diagnostic messages
#' @param nn nearest-neighbors computed by \code{\link{findNearestNeighbors}}
#' @param radius Integer: Indices that dist(targetInd - queryInd) <= radius
#' @param  adjust.method method to adjust p-value,
#'  passed to \code{\link{topTable}}
#' @param ... additional arguments to pass to \code{\link[parallel]{mclapply}}

#' @return A list of output images, including the moderated T map, p-value map,
#' adjusted p-value map, and the pseudo T map.
#' @export


nifti_local_moderated_t_stat = function(imgs, mask = NULL, nn, radius, adjust.method = "BH", verbose = FALSE, ...){

  maps = nifti_images_to_matrix(imgs = imgs, mask = mask)
  nrow.map = nrow(maps)

  t.stat.list = parallel::mclapply(1:nrow.map,
                     function(center){
                       # require(limma)
                       idx = nn$index[center, nn$distance[center,]<= radius]
                       fit = limma::lmFit(maps[idx,])
                       eb.fit = limma::eBayes(fit)
                       tstat = eb.fit$t[1]
                       pval = eb.fit$p.value[1]
                       # local.eb.fit = sapply(eb.fit, "[", 1)
                       adj.pval = limma::topTable(eb.fit, sort.by = 'none', adjust.method = adjust.method, number = length(idx))$adj.P.Val[1]
                       psuedo.t = eb.fit$coefficients[1]/sqrt(mean((eb.fit$sigma)^2))/eb.fit$stdev.unscaled[1]
                       return(c(tstat, pval, adj.pval, psuedo.t))
                     })

  t.stat = do.call("cbind", t.stat.list)

  tmap.limma = mask * NA
  pvalmap.limma = mask * NA
  adj.pvalmap.limma = mask * NA
  pseudo.tmap.limma = mask * NA
  if (is.null(mask)){
    tmap.limma = t.stat[1,]
    pvalmap.limma = t.stat[2,]
    adj.pvalmap.limma = t.stat[3,]
    pseudo.tmap.limma = t.stat[4,]
  }else{
    tmap.limma[mask>0] = t.stat[1,]
    pvalmap.limma[mask>0] = t.stat[2,]
    adj.pvalmap.limma[mask>0] = t.stat[3,]
    pseudo.tmap.limma[mask>0] = t.stat[4,]
  }
  return(list(tmap = tmap.limma,
              pvalmap = pvalmap.limma,
              adj.pvalmap = adj.pvalmap.limma,
              pseudo.tmap = pseudo.tmap.limma)
         )

}



R/nifti_eBayes.R

#' Run `eBayes` on NIfTI Images
#'
#' @inheritParams nifti_lmFit
#' @param proportion numeric value between 0 and 1, assumed proportion of
#' voxels which are differentially expressed, passed to \code{\link{eBayes}}
#' @param stdev.coef.lim numeric vector of length 2, assumed
#' lower and upper limits for the standard deviation of
#' log2-fold-changes for differentially expressed voxels,
#' passed to \code{\link{eBayes}}
#' @param trend logical, should an intensity-trend be allowed for the
#' prior variance? Default is that the prior variance is constant,
#' passed to \code{\link{eBayes}}
#' @param robust logical, should the estimation of df.prior and
#' var.prior be robustified against outlier sample variances,
#' passed to \code{\link{eBayes}}
#' @param winsor.tail.p numeric vector of length 1 or 2,
#' giving left and right tail proportions of x to Winsorize.
#' Used only when \code{robust=TRUE}, passed to \code{\link{eBayes}}
#' @param adjust.method method to adjust p-value,
#'  passed to \code{\link{topTable}}
#' @param coef column number or column name specifying which
#' coefficient or contrast of the linear model is of interest,
#' passed to \code{\link{topTable}}
#' @return
#' @export
#'
#' @return A list of output images from `eBayes` and the linear fit
#' @examples
#' tarfile = system.file("extdata", "can.tar.gz", package = "limmi")
#' tarfile
#' exdir = tempfile()
#' dir.create(exdir)
#' files = untar(tarfile = tarfile, list = TRUE, exdir = exdir)
#' files = files[!grepl("^[.]", basename(files))]
#' unz = untar(tarfile = tarfile, files = files, exdir = exdir)
#' files = files[ grepl("hdr", basename(files))]
#' files
#' files = file.path(exdir, files)
#' res = nifti_eBayes(files, verbose = FALSE)
#' \dontrun{
#' design_mat = cbind(x1 = rep(c(0, 1), each = 6),
#' x2 = rep(c(0, 1, 0, 1), each = 3))
#' res2 = nifti_eBayes(files, design = design_mat, coef = 2)
#' }
nifti_eBayes = function(
  imgs, mask = NULL, verbose = TRUE,
  ...,
  proportion = 0.01, stdev.coef.lim = c(0.1,4),
  trend = FALSE, robust = FALSE, winsor.tail.p = c(0.05,0.1),
  coef = NULL,
  adjust.method = "BH") {
  if (is.factor(imgs)) {
    imgs = as.character(imgs)
  }
  img = RNifti::asNifti(imgs[[1]])
  if (!is.null(mask)) {
    mask =  RNifti::asNifti(mask) > 0
  } else {
    mask = array(TRUE, dim = dim(img))
  }

  if (verbose) {
    msg = paste0("Combining Images into a Matrix")
    message(msg)
  }
  mat = nifti_images_to_matrix(imgs, mask, verbose = verbose)
  # n = nrow(mat)
  if (verbose) {
    msg = paste0("Making lmFit")
    message(msg)
  }
  fit = limma::lmFit(mat, ...)

  if (verbose) {
    msg = paste0("Running eBayes")
    message(msg)
  }
  mask = RNifti::asNifti(mask, reference = img)
  L = eBayes_to_images(
    fit,
    mask = mask,
    verbose = verbose,
    proportion = proportion,
    stdev.coef.lim = stdev.coef.lim,
    trend = trend,
    robust = robust,
    winsor.tail.p = winsor.tail.p,
    coef = coef,
    adjust.method = adjust.method
  )

  L$lm_fit = fit
  L$mask = mask
  L
}

#' @rdname nifti_eBayes
#' @param fit an MArrayLM fitted model object produced by `lmFit` or
#' `contrasts.fit`.
#' @export
eBayes_to_images = function(
  fit,
  mask, verbose = TRUE,
  proportion = 0.01, stdev.coef.lim = c(0.1,4),
  trend = FALSE, robust = FALSE, winsor.tail.p = c(0.05,0.1),
  coef = NULL,
  adjust.method = "BH") {
  cols_to_grab = c("coefficients", "stdev.unscaled",
                   "t", "p.value", "lods", "adjusted_p_value")
  if (verbose) {
    msg = paste0("Running p-value adjustment")
    message(msg)
  }
  n = nrow(fit$coefficients)

  eb.fit = limma::eBayes(
    fit,
    proportion = proportion,
    stdev.coef.lim = stdev.coef.lim,
    trend = trend, robust = robust,
    winsor.tail.p = winsor.tail.p)

  nc = ncol(eb.fit$coefficients)
  eb.fit$adjusted_p_value = sapply(
    seq(nc),
    function(x) {
      limma::topTable(
        eb.fit,
        sort.by = "none",
        coef = x,
        adjust.method = adjust.method,
        number = n)$adj.P.Val
    })
  colnames(eb.fit$adjusted_p_value) = colnames(fit$coefficients)

  if (!is.null(coef)) {
    eb.fit$coef_adjusted_p_value = matrix(
      limma::topTable(
        eb.fit,
        sort.by = "none",
        coef = coef,
        adjust.method = adjust.method,
        number = n)$adj.P.Val, ncol = 1)
    cols_to_grab = c(cols_to_grab, "coef_adjusted_p_value")
  }

  if (verbose) {
    msg = paste0("Transforming to images")
    message(msg)
  }
  img = mask
  out_images = lapply(cols_to_grab, function(x) {
    sub_eb = eb.fit[[x]]
    if (verbose > 1) {
      message(x)
    }
    res = apply(sub_eb, 2, function(r) {
      list(remake_nifti_image(r, img = img, mask = mask))
    })
    res = lapply(res, function(ll) ll[[1]])
    if (length(res) == 1) {
      res = res[[1]]
    }
    res
  })
  names(out_images) = cols_to_grab
  L = list(
    images = out_images,
    empirical_bayes = eb.fit,
    adjust.method = adjust.method
  )
  return(L)
}
